### PR TITLE
Add CUDA support for vectorization.

### PR DESCRIPTION
**Pull request by stefanseefeld**
This patch contains a test that relies on a previously committed pull request (#1404).

### CODE REVIEW

File: numba/npyufunc/__init__.py [hunk 1]
```diff
@@ -6,3 +6,22 @@
...
+
+    def init_vectorize():
+        from numba.cuda.vectorizers import CUDAVectorize
+        return CUDAVectorize
+
+    def init_guvectorize():
+        from numba.cuda.vectorizers import CUDAGUFuncVectorize
+        return CUDAGUFuncVectorize
+
+    Vectorize.target_registry.ondemand['gpu'] = init_vectorize
```
Comment (gmarkall) [inline numba/npyufunc/__init__.py]: Whilst having a target called 'gpu' before probably was OK, I think as this API is new to Numba, we should probably not refer to it as 'gpu', in order to not cause confusion if and when ufuncs on non-CUDA GPU-like architectures are supported. (In other words this line and the other one referencing 'gpu' should probably be removed).

**Comment by gmarkall** [comment]
I certainly can't claim to understand all this code, but I think the organisation of it looks OK. 

One thing that would be handy is if there were a small test of `CudaKernelBase.forall`, because we will be relying on that API. This could just test with a simple kernel like:

```
@cuda.jit
def cuda_arange(ary, count):
    i = cuda.grid(1)
    if i < count:
        ary[i] = i
```

and the test would ensure that calling `cuda_arange.forall(len(arr))(arr, len(arr))` sets the values in the array such that it is equal to the result of `np.arange(len(arr), arr.dtype)` (as an example).

File: numba/cuda/cudadrv/autotune.py [hunk 2]
```diff
@@ -0,0 +1,296 @@
...
+- Compute warp occupany histogram
+"""
+from __future__ import division, absolute_import, print_function
+import math
+import re
+
+SMEM16K = 16 * 2 ** 10
+SMEM48K = 48 * 2 ** 10
+SMEM64K = 64 * 2 ** 10
+SMEM96K = 64 * 2 ** 10
```
Comment (pitrou) [inline numba/cuda/cudadrv/autotune.py]: Those definitions look fishy.

File: numba/tests/npyufunc/test_cuda_gufunc.py [hunk 3]
```diff
@@ -0,0 +1,190 @@
+from __future__ import print_function, absolute_import
```
Comment (pitrou) [inline numba/tests/npyufunc/test_cuda_gufunc.py]: I'm thinking about something... if these are all CUDA tests, shouldn't they be in `numba/cuda/tests/cudapy/`?

Comment (stefanseefeld) [inline numba/tests/npyufunc/test_cuda_gufunc.py]: There are certainly multiple axes to organize them. I have grouped them together as ufunc tests.

Comment (pitrou) [inline numba/tests/npyufunc/test_cuda_gufunc.py]: You could still group them as ufunc tests inside the CUDA tests directory ;) The separation between CPU and GPU tests right now is very readable, it becomes less so if we start putting CUDA tests in the main tests directory.

Comment (stefanseefeld) [inline numba/tests/npyufunc/test_cuda_gufunc.py]: Fair enough. I'll move them tomorrow.

File: numba/npyufunc/__init__.py (hunk 1 above)
Comment (stefanseefeld) [inline numba/npyufunc/__init__.py]: OK, That's fixed.

File: numba/cuda/cudadrv/autotune.py (hunk 2 above)
Comment (stefanseefeld) [inline numba/cuda/cudadrv/autotune.py]: That's fixed.

File: docs/source/cuda/ufunc.rst [hunk 4]
```diff
@@ -0,0 +1,154 @@
...
+
+    from numba import guvectorize
+
+    @guvectorize(..., target='cuda')
+    def very_complex_kernel(A, B, C):
+        ...
+
+    very_complex_kernel.max_blocksize = 32  # limits to 32 threads per block
+
+.. comment
```
Comment (pitrou) [inline docs/source/cuda/ufunc.rst]: Does ".. comment" mean this piece of doc is commented out?

File: numba/cuda/compiler.py [hunk 5]
```diff
@@ -477,6 +521,27 @@ def _prepare_args(self, ty, val, stream, retr, kernelargs):
...
+        else:
+            # Get CUDA Function
+            cufunc = self._func.get()
+            at = AutoTuner(info=cufunc.attrs, cc=cufunc.device.compute_capability)
+            self._autotune = at
+            return self._autotune
+
+    @property
+    def occupancy(self):
+        """calculate the theoretical occupancy of the kernel given the
```
Comment (pitrou) [inline numba/cuda/compiler.py]: What does it return? A number of threads?

File: numba/cuda/compiler.py [hunk 6]
```diff
@@ -477,6 +521,27 @@ def _prepare_args(self, ty, val, stream, retr, kernelargs):
             raise NotImplementedError(ty, val)
 
 
+    @property
+    def autotune(self):
```
Comment (pitrou) [inline numba/cuda/compiler.py]: Perhaps add a docstring here?

File: numba/cuda/deviceufunc.py [hunk 7]
```diff
@@ -0,0 +1,742 @@
...
+from numba.npyufunc.ufuncbuilder import _BaseUFuncBuilder
+from numba import sigutils, types
+from numba.typing import signature
+from numba.npyufunc.sigparse import parse_signature
+
+if IS_PY3:
+    def _exec(codestr, glbls):
+        exec(codestr, glbls)
+else:
+    eval(compile("""
```
Comment (pitrou) [inline numba/cuda/deviceufunc.py]: Note we generally use the six library for such things:
http://pythonhosted.org/six/#syntax-compatibility

File: docs/source/cuda/ufunc.rst (hunk 4 above)
Comment (stefanseefeld) [inline docs/source/cuda/ufunc.rst]: Yes. (http://docutils.sourceforge.net/docs/user/rst/quickref.html#comments)

File: numba/cuda/compiler.py (hunk 6 above)
Comment (stefanseefeld) [inline numba/cuda/compiler.py]: Done. (It's minimal as it isn't clear to what degree the API should be considered 'public'.

**Pull request merged**